src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py

```diff
@@ -15,31 +15,94 @@
 """This script can be used to convert checkpoints provided in the `mamba2_ssm` library into the format provided in HuggingFace `transformers`. It depends on the `mamba2_ssm` package to be installed."""
 
 import argparse
+import json
+from os import path
+from typing import Dict, Optional, Tuple
 
 import torch
 from safetensors import safe_open
+from safetensors.torch import save_model
 
-from transformers import LlamaTokenizerFast, Mamba2Config, Mamba2ForCausalLM
+from transformers import AutoTokenizer, LlamaTokenizerFast, Mamba2Config, Mamba2ForCausalLM
 
 
-def convert_mamba2_checkpoint_file_to_huggingface_model_file(
-    mamba2_checkpoint_path: str, tokenizer_model_path: str, output_dir: str
-) -> None:
+def convert_ssm_config_to_hf_config(config_ssm: Dict) -> Tuple[Mamba2Config, bool]:
+    """Convert a Mamba2Config from mamba_ssm to a Mamba2Config from here."""
     hf_config = Mamba2Config()
-    hf_model = Mamba2ForCausalLM(hf_config)
+
+    # Flag for codestral model
+    is_not_codestral = "dim" not in config_ssm
+
+    # Set important values from config and recalculate other resulting entries
+    hf_config.hidden_size = config_ssm["d_model"] if is_not_codestral else config_ssm["dim"]
+    hf_config.num_heads = (hf_config.hidden_size * hf_config.expand) // hf_config.head_dim
+    hf_config.num_hidden_layers = config_ssm["n_layer"] if is_not_codestral else config_ssm["n_layers"]
+    hf_config.n_groups = config_ssm.get("n_groups", 1)
+    hf_config.residual_in_fp32 = config_ssm["residual_in_fp32"]
+    hf_config.tie_word_embeddings = config_ssm["tie_embeddings"]
+    hf_config.norm_before_gate = False
```

Comment:
This might've been a `False` default value: https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/modules/mamba2.py#L51

It's `True` as default in the config.

Reply:
Ohh nice catch! Indeed, I just tested and got a different result in the code - I believe in that case that `MambaRMSNormGated` should be updated in a different PR since only one behaviour is supported right now, RMSNorm(x) * F.silu(z) instead of RMSNorm(x * F.silu(z))

Reply:
I think the current implementation follows `norm_before_gate` = `False` https://github.com/huggingface/transformers/blob/2a5a6ad18aa22e98429bb5ecb880660328030ea0/src/transformers/models/mamba2/modeling_mamba2.py#L182-L185

The original also does `RMSNorm(x * F.silu(z)`
https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/ops/triton/layernorm_gated.py#L26-L27

It will probably only be erroneous when training with the fast path on the fused kernel. But yes, a respective PR may be appropriate :D

Reply:
Has been handled by #32686